posts/spring2021/js-chats-2/index.md

```diff
@@ -0,0 +1,176 @@
+---
+date: 2021-04-11
+title: Node.js Profiling
+subtitle: JavaScript Chats Hack Session 2 Spring 2021 
+description: >
+  The first step of optimization is to get a hold of performance. Using the
+  Chrome DevTools, we profile a Node.js application. 
+---
+
+
+In the previous [JS Chats on optimization](/posts/fall2019/js-chats-4), we took
+a peek at the Chrome DevTools and how we can use the Performance tools to
+profile our web application loading time. Since both Node.js and Chromium both
+rely on the V8 engine, the DevTools in Chrome can actually be a debugger and a
+profiler for Node.js as well! Let's see how we can use it. 
+
+Here, we built a very simple sever program, with one API endpoint to get some
+cryptocurrency prices.  
+
+```js
+app.get('/coin-prices', async (req, res) => {
+    const listOfCoins = ['dogecoin', 'bitcoin', 'ethereum', 'polkadot',
+        'litecoin', 'cardano', 'bitcoin-cash', 'stellar', 'chainlink',
+        'binancecoin'];
+
+    const prices = {};
+    for (const coinId of listOfCoins) {
+        prices[coinId] = await getCoinPrice(coinId);
+    }
+    res.json(prices);
+});
+```
+
+The full server program can be found here:
+> https://github.com/uclaacm/js-chats-s21/tree/main/node-profile
+
+The endpoint uses a list of coin IDs, queries their prices against an external
+API `getCoinPrice`, and return them as a JSON object. For those who has worked
+with async programming in JavaScript before, you might have already noticed the
+bad performance pattern. If not, don't worry. We will show where the bad spot
+is and how to optimize it. Let's try to profile this.
+
+To allow for profiling, we invoke the server program with the `--inspect` flag.
+
+```bash
+node --inspect index.js
+```
+
+The terminal outputs the following
+
+```
+Debugger listening on ws://127.0.0.1:9229/9b78b923-925f-462c-8dca-345b3b6f9f38
+For help, see: https://nodejs.org/en/docs/inspector
+```
+
+This means the profiler and debugger has been invoked. Now, we head to Chrome
+and open DevTools (right click anywhere and click inspect). On the top left corner, 
+a Node.js icon is going to pop up. 
+
+![button to open Node.js debugger in Chrome DevTools](images/devtool-node.jpg)
+
+After they click on it, the Node.js profiler will pop up. 
+
+![Node.js Chrome debugger](images/node-chrome-profiler.jpg)
```

Comment:
Why are these JPEGs rather than PNGs?

Reply:
because me dumb i forget to change file saving format on my computer